File: Common/Verify.vb
Namespace Common

    Public Class Verify

        <DebuggerHidden()>
        Public Shared Sub NonNullArg(Of T As Class)(ByVal O As T, Optional ByVal Name As String = Nothing, Optional ByVal Message As String = Nothing)
            If O Is Nothing Then
                If Name Is Nothing Then
                    Throw New ArgumentNullException()
                ElseIf Message Is Nothing Then
                    Throw New ArgumentNullException(Name)
                Else
                    Throw New ArgumentNullException(Name, Message)
                End If
            End If
        End Sub

        <DebuggerHidden()>
        Public Shared Sub TrueArg(ByVal T As Boolean, Optional ByVal Name As String = Nothing, Optional ByVal Message As String = Nothing)
            If Not T Then
                If Name Is Nothing And Message Is Nothing Then
                    Throw New ArgumentException()
                Else
                    Throw New ArgumentException(Message, Name)
                End If
            End If
        End Sub

        <DebuggerHidden()>
        Public Shared Sub RangeArg(Of T As IComparable(Of T))(ByVal Start As T, ByVal V As T, ByVal [End] As T, Optional ByVal Name As String = Nothing, Optional ByVal Message As String = Nothing)
            If Start.CompareTo(V) > 0 OrElse V.CompareTo([End]) > 0 Then
                If Name Is Nothing Then
                    Throw New ArgumentOutOfRangeException()
                ElseIf Message Is Nothing Then
                    Throw New ArgumentOutOfRangeException(Name, V, String.Format("Argument must be between '{0}' and '{1}'.", Start, [End]))
                Else
                    Throw New ArgumentOutOfRangeException(Name, V, Message)
                End If
            End If
        End Sub

        <DebuggerHidden()>
        Public Shared Sub FalseArg(ByVal T As Boolean, Optional ByVal Name As String = Nothing, Optional ByVal Message As String = Nothing)
            If T Then
                If Name Is Nothing And Message Is Nothing Then
                    Throw New ArgumentException()
                Else
                    Throw New ArgumentException(Message, Name)
                End If
            End If
        End Sub

        <DebuggerHidden()>
        Public Shared Sub NonNull(Of T As Class)(ByVal O As T, Optional ByVal Name As String = Nothing)
            If O Is Nothing Then
                If Name Is Nothing Then
                    Throw New NullReferenceException(String.Format("Object reference '{0}' not set to an instance of an object.", Name))
                Else
                    Throw New NullReferenceException()
                End If
            End If
        End Sub

        <DebuggerHidden()>
        Public Shared Sub [True](ByVal T As Boolean, Optional ByVal Message As String = Nothing)
            If Not T Then
                Throw New InvalidOperationException(Message)
            End If
        End Sub

        <DebuggerHidden()>
        Public Shared Sub [False](ByVal T As Boolean, Optional ByVal Message As String = Nothing)
            If T Then
                Throw New InvalidOperationException(Message)
            End If
        End Sub

        <DebuggerHidden()>
        Public Shared Sub Fail(Optional ByVal Message As String = Nothing)
            Throw New InvalidOperationException(Message)
        End Sub

        <DebuggerHidden()>
        Public Shared Sub FailArg(Optional ByVal Name As String = Nothing, Optional ByVal Message As String = Nothing)
            If Name Is Nothing And Message Is Nothing Then
                Throw New ArgumentException()
            Else
                Throw New ArgumentException(Message, Name)
            End If
        End Sub

    End Class

End Namespace

File: Common/Verify.cs
using System;
using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;

namespace Ks.Common
{
#pragma warning disable CA2208 // Instantiate argument exceptions correctly
    public class Verify
    {
        [DebuggerHidden()]
        public static void NonNullArg<T>([NotNull] T O, string Name = null, string Message = null) where T : class
        {
            if (O == null)
            {
                if (Name == null)
                {
                    throw new ArgumentNullException();
                }
                else if (Message == null)
                {
                    throw new ArgumentNullException(Name);
                }
                else
                {
                    throw new ArgumentNullException(Name, Message);
                }
            }
        }

        [DebuggerHidden()]
        public static void TrueArg([DoesNotReturnIf(false)] bool T, string Name = null, string Message = null)
        {
            if (!T)
            {
                if (Name == null & Message == null)
                {
                    throw new ArgumentException();
                }
                else
                {
                    throw new ArgumentException(Message, Name);
                }
            }
        }

        [DebuggerHidden()]
        public static void RangeArg<T>(T Start, T V, T End, string Name = null, string Message = null) where T : IComparable<T>
        {
            if (Start.CompareTo(V) > 0 || V.CompareTo(End) > 0)
            {
                if (Name == null)
                {
                    throw new ArgumentOutOfRangeException();
                }
                else if (Message == null)
                {
                    throw new ArgumentOutOfRangeException(Name, V, string.Format("Argument must be between '{0}' and '{1}'.", Start, End));
                }
                else
                {
                    throw new ArgumentOutOfRangeException(Name, V, Message);
                }
            }
        }

        [DebuggerHidden()]
        public static void FalseArg([DoesNotReturnIf(true)] bool T, string Name = null, string Message = null)
        {
            if (T)
            {
                if (Name == null & Message == null)
                {
                    throw new ArgumentException();
                }
                else
                {
                    throw new ArgumentException(Message, Name);
                }
            }
        }

        [DebuggerHidden()]
        public static void NonNull<T>([NotNull] T O, string Name = null) where T : class
        {
            if (O == null)
            {
                if (Name == null)
                {
                    throw new NullReferenceException(string.Format("Object reference '{0}' not set to an instance of an object.", Name));
                }
                else
                {
                    throw new NullReferenceException();
                }
            }
        }

        [DebuggerHidden()]
        public static void True([DoesNotReturnIf(false)] bool T, string Message = null)
        {
            if (!T)
            {
                throw new InvalidOperationException(Message);
            }
        }

        [DebuggerHidden()]
        public static void False([DoesNotReturnIf(true)] bool T, string Message = null)
        {
            if (T)
            {
                throw new InvalidOperationException(Message);
            }
        }

        [DebuggerHidden()]
        [DoesNotReturn]
        public static InvalidOperationException Fail(string Message = null)
        {
            throw new InvalidOperationException(Message);
        }

        [DebuggerHidden()]
        [DoesNotReturn]
        public static ArgumentException FailArg(string Name = null, string Message = null)
        {
            if (Name == null & Message == null)
            {
                throw new ArgumentException();
            }
            else
            {
                throw new ArgumentException(Message, Name);
            }
        }
    }
}
